### PR TITLE
Project-aware Chat Agent

### PR DESCRIPTION
**Pull request by CamilleLetavernier**
#### What it does

This PR provides a new Agent (TheiaWorkspaceAgent), which is aware of the project contents. It has access to the contents of the workspace via two dedicated tools:

- getProjectFileList(): return the list of files available in the project
- getFileContent(filePath: string): return the content of the file

Most notably, this PR includes the following changes:

- ToolService, to register tools (currently on the frontend) that can be invoked by the backend (when the LLM Provider is defined)
- ~~The TheiaWorkspaceAgent is currently set as the default agent, because we don't have a way to pick specific agents (yet!)~~
  - ~~This can be reverted once https://github.com/eclipsesource/theia/pull/121 is merged~~
  - The above PR has been merged, so the Workspace agent needs to be queried with `@TheiaWorkspaceAgent`
- ChatMessages can now specify the type "System", in addition to "AI" and "user", to define the system prompt
- DefaultChatAgent was slightly refactored to make it easier to extend (especially to inject a system prompt from a custom template)
- Log the full message stack (Actor, Type and Query) in the AI Chat History

#### How to test

- open a workspace with a project (all workspace contents will be considered by the agent)
- ask the agent some information about the project (e.g. How do I build this project, or What is this project about?)
- **Important**: Prefix **all** your queries with `@TheiaWorkspaceAgent`, otherwise you'll fall back to the Default Agent which doesn't have access to the workspace

#### Follow-ups

https://github.com/eclipsesource/osweek-2024/issues/76 and https://github.com/eclipsesource/osweek-2024/issues/77

### CODE REVIEW

File: packages/ai-openai/src/node/openai-model-provider.ts [hunk 1]
```diff
@@ -46,24 +51,74 @@ export class OpenAiModel implements LanguageModel {
...
+        //     stream: {
+        //         [Symbol.asyncIterator](): AsyncIterator<LanguageModelStreamResponsePart> {
+        //             return {
+        //                 next(): Promise<IteratorResult<LanguageModelStreamResponsePart>> {
+        //                     return stream1[Symbol.asyncIterator]().next().then(chunk => chunk.done ? chunk : { value: chunk.value.choices[0]?.delta, done: false });
+        //                 }
+        //             };
+        //         }
+        //     }
+        // };
```
Comment (sgraband) [inline packages/ai-openai/src/node/openai-model-provider.ts:93]: Can we remove this?

File: packages/ai-core/src/common/language-model-tool-service.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
+// *****************************************************************************
+// import { ToolRequest } from './language-model';
```
Comment (sgraband) [inline packages/ai-core/src/common/language-model-tool-service.ts:16]:
```suggestion
```

File: packages/ai-core/src/browser/frontend-language-model-registry.ts [hunk 3]
```diff
@@ -87,6 +88,9 @@ export class FrontendLanguageModelRegistryImpl
 
     @inject(AISettingsService)
     protected settingsService: AISettingsService;
+    
+    @inject(LanguageModelToolServer)
+    protected toolServer: LanguageModelToolServer;
```
Comment (sgraband) [inline packages/ai-core/src/browser/frontend-language-model-registry.ts:93]: Not needed i believe

File: packages/ai-chat/src/common/chat-service.ts [hunk 4]
```diff
@@ -143,7 +143,9 @@ export class ChatServiceImpl implements ChatService {
         const chatAgents = this.chatAgentService.getAgents();
         if (chatAgents.length > 0) {
             // TODO collect the correct agent
-            this.chatAgentService.invokeAgent(chatAgents[0].id, requestModel);
+            // FIXME: This is a temporary solution to get the TheiaWorkspaceAgent (Issue #6)
+            const agent = this.chatAgentService.getAgent('TheiaWorkspaceAgent') ?? chatAgents[0];
+            this.chatAgentService.invokeAgent(agent.id, requestModel);
```
Comment (sgraband) [inline packages/ai-chat/src/common/chat-service.ts]: Should be fixed before merging

Comment (sgraband) [reply packages/ai-chat/src/common/chat-service.ts]: No longer required

File: packages/ai-core/src/browser/frontend-language-model-registry.ts (hunk 3 above)
Comment (CamilleLetavernier) [reply packages/ai-core/src/browser/frontend-language-model-registry.ts:93]: It's a bit tricky: the client doesn't need the server, but we still need to initialize the binding, so the client can register itself to the server. This is dirty way to achieve this.

Comment (CamilleLetavernier) [reply packages/ai-core/src/browser/frontend-language-model-registry.ts:96]: I added a comment to clarify this

**Review by sgraband — APPROVED** [review]

**Pull request merged**